build/dangerjs/detectNonApprovedDependencies/utils/checkPackageVersionConstraints.ts

```diff
@@ -0,0 +1,62 @@
+import * as semver from 'semver'
+
+import approvedPackages, { isIgnored } from '../approvedPackages'
+import { getPackageName, getPackageVersion } from './packageNameUtils'
+
+export type FailedConstraintsExplanation = {
+  approvedPackages: string[]
+  failedConstraints: string[]
+}
+
+export type ApprovalCheckResult = {
+  isApproved: boolean
+  explain?: FailedConstraintsExplanation
+}
+
+const checkPackageVersionConstraints = (
+  packageId: string,
+  versionConstraints: string[],
+): ApprovalCheckResult => {
+  const packageName = getPackageName(packageId)
+
+  if (isIgnored(packageId)) {
+    return { isApproved: true }
+  }
+
+  const versionConstraintsCheckResult = versionConstraints.reduce(
+    (acc, versionConstraint) => ({ ...acc, [versionConstraint]: false }),
+    {},
+  )
+
+  const approvedPackageIds = approvedPackages.filter(item => item.startsWith(`${packageName}@`))
+
+  approvedPackageIds.map(getPackageVersion).forEach(approvedPackageVersion => {
+    Object.keys(versionConstraintsCheckResult).forEach(versionConstraint => {
+      if (semver.satisfies(approvedPackageVersion, versionConstraint)) {
+        versionConstraintsCheckResult[versionConstraint] = true
+      }
+    })
+  })
+
+  const failedConstraints = Object.keys(versionConstraintsCheckResult).reduce(
+    (acc, versionConstraint) =>
+      !versionConstraintsCheckResult[versionConstraint] ? [...acc, versionConstraint] : acc,
+    [],
+  )
```

Comment:
Can use `lodash` in this case? It's hard to understand what happens here...

Reply:
have discussed offline, this `reduce` expression will be removed, overall code of the module will be simplified